### PR TITLE
Maxim/clients

### PR DESCRIPTION
**Pull request by Druddigon**
Users can now choose a username for the chat. Chat now shows a list of connected users. (Pushing 1 commit tomorrow morning with fixes as per Kevin's comments)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -17,5 +17,9 @@
         "gulp-notify": "latest",
         "gulp-rename": "latest"
     },
-    "author": "developer"
+    "author": "developer",
+    "scripts": {
+      "start": "node app.js",
+      "postinstall": "gulp default"
```
Comment (rongduan-zhu) [inline package.json]: ```gulp```

File: public/index.html [hunk 2]
```diff
@@ -3,30 +3,54 @@
...
+  <link rel="stylesheet" type="text/css" href="/css/app.css">
 </head>
 <body>
-  <div ng-controller="MessageCtrl as messageCtrl" ng-cloak>
-    <div class="content" >
-      <div ng-repeat="message in messageCtrl.messages.data">
-        <div>{{ ::message.message }}</div>
+  <div ng-controller="ClientCtrl as clientCtrl" ng-cloak>
+    <div ng-controller="MessageCtrl as messageCtrl" ng-show="clientCtrl.clientId">
+      <!-- BEGIN: Message list -->
```
Comment (rongduan-zhu) [inline public/index.html]: don't think you need comments here. your code is very self explanatory.

File: public/index.html [hunk 3]
```diff
@@ -3,30 +3,54 @@
...
+      <div class="content">
+        <div ng-repeat="message in messageCtrl.messages.data">
+          <div>{{ ::message.message }}</div>
+        </div>
       </div>
+      <!-- END: Message list -->
+
```
Comment (rongduan-zhu) [inline public/index.html]: same as the newline here. not really necessary

File: public/index.html [hunk 4]
```diff
@@ -3,30 +3,54 @@
...
+      <!-- BEGIN: User list -->
+      <div class="user-list" ng-controller="UserCtrl as userCtrl">
+        <p>Users:</p>
+        <ul ng-repeat="(userId, username) in userCtrl.userData.userList">
+          <li>{{ username }}</li>
+        </ul>
+      </div>
+      <!-- END: User list -->
+
```
Comment (rongduan-zhu) [inline public/index.html]: same here

File: statics/js/client.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
+(function() {
+  chat.factory('Connection', function(socket, $rootScope) {
+    'use strict';
```
Comment (rongduan-zhu) [inline statics/js/client.js]: why not just ```'use strict'``` at the top?

File: statics/js/message.js [hunk 6]
```diff
@@ -1,10 +1,11 @@
 chat.factory('Messages', function($rootScope, socket, _) {
```
Comment (rongduan-zhu) [inline statics/js/message.js]: ```_``` no longer needs to be injected. i don't think it is used.

File: statics/js/users.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
+(function() {
+  chat.factory('User', function(socket, $rootScope) {
+    'use strict';
```
Comment (rongduan-zhu) [inline statics/js/users.js]: same here, i think you can move it to the top of the file

File: statics/js/users.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
+(function() {
+  chat.factory('User', function(socket, $rootScope) {
```
Comment (rongduan-zhu) [inline statics/js/users.js]: Maybe call this ```Users```? because the messages factory is called ```Messages```.

File: app.js [hunk 9]
```diff
@@ -1,32 +1,52 @@
...
 app.get('/', function(request, response) {
   response.sendFile(__dirname + '/index.html');
 });
 
 var io = require('socket.io')(http)
 
+// Store chat users with key = user ID and value = username
+var chatUserList = {};
 
 io.sockets.on('connection', function (socket) {
```
Comment (rongduan-zhu) [inline app.js]: why not just ```io.on```?

File: app.js [hunk 10]
```diff
@@ -1,32 +1,52 @@
...
   console.log('A user has connected');
 
+  // User has joined chat, add their ID and username to user list
+  socket.on('join-chat', function(data) {
+    console.log('a user with name ' + data.username + ' has joined the chat.');
+    chatUserList[socket.id] = data.username;
+
+    // Notify client of their successful joining of chat
+    io.to(socket.id).emit('new-client', { id: socket.id, username: data.username });
```
Comment (rongduan-zhu) [inline app.js]: how come not just ```socket.emit``` since they produce the same effect and the latter is shorter.

File: package.json (hunk 1 above)
Comment (rongduan-zhu) [inline package.json]: just ```gulp``` is fine

File: statics/js/message.js (hunk 6 above)
Comment (Druddigon) [inline statics/js/message.js]: Removed the ```_```

File: app.js (hunk 9 above)
Comment (Druddigon) [inline app.js]: Didn't realise they were the same thing, I'll change it to the shorter form.

File: app.js (hunk 10 above)
Comment (Druddigon) [inline app.js]: In reference to line 29?

File: package.json (hunk 1 above)
Comment (Druddigon) [inline package.json]: Changed this

File: app.js (hunk 10 above)
Comment (rongduan-zhu) [inline app.js]: I thought if you want to send to all clients you do ```io.sockets.emit```, where as ```socket.emit``` is just sending to that particular person who just connected.

File: package.json [hunk 11]
```diff
@@ -17,5 +15,13 @@
         "gulp-notify": "latest",
         "gulp-rename": "latest"
     },
-    "author": "developer"
+    "author": "developer",
+    "scripts": {
+      "start": "node app.js",
+<<<<<<< HEAD
```
Comment (rongduan-zhu) [inline package.json]: mate....you still drunk from last night?

**Comment by rongduan-zhu** [comment]
:+1:

**Pull request merged**